### PR TITLE
Added an article on S2S auth

### PR DESCRIPTION
**Pull request by ksimons**

### CODE REVIEW

File: resources/articles/s2s-auth.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+## See also
+
+- [Obtaining tokens](/oauth/token/)
+- [Token introspection](/token-introspection/)
+
+:body
+
+In addition to the user auth facilities provided by the platform, SPiD also provides a mechanism for
+when two services need to communicate directly without a user present. In these situations, a service
```
Comment (zamzterz) [inline resources/articles/s2s-auth.md]: "when two services need to communicate directly without a user present" -> "when two services want to communicate directly with each other" (The general case of protected resources doesn't even care about user existence, so user presence is not a factor here :)).

File: resources/articles/s2s-auth.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+needs to authenticate and authorize the other services which access its protected resources. We refer
+to this as service-to-service (henceforth S2S) auth.
+
+S2S auth in SPiD is made up of three basic steps:
+1. A client [obtains a client (a.k.a. server) token](/oauth/token/) via
```
Comment (zamzterz) [inline resources/articles/s2s-auth.md]: Does it suffice to only name it "client token" (to minimize confusion)?

File: resources/articles/s2s-auth.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+- [Token introspection](/token-introspection/)
+
+:body
+
+In addition to the user auth facilities provided by the platform, SPiD also provides a mechanism for
+when two services need to communicate directly without a user present. In these situations, a service
+needs to authenticate and authorize the other services which access its protected resources. We refer
+to this as service-to-service (henceforth S2S) auth.
+
+S2S auth in SPiD is made up of three basic steps:
```
Comment (zamzterz) [inline resources/articles/s2s-auth.md]: Missing newline after this for the list to be rendered correctly.

File: resources/articles/s2s-auth.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+1. A client [obtains a client (a.k.a. server) token](/oauth/token/) via
+a `client_credentials` grant.
+1. The client makes a request to the resource server passing the newly acquired token as a
```
Comment (zamzterz) [inline resources/articles/s2s-auth.md]: Missing definition of resource server, suggestion: "The client makes a request to the resource server" -> 'The client makes a request to the server hosting the protected resource, known as the "resource server", passing [...]'

File: resources/articles/s2s-auth.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+![Server-to-server communication flow](/images/server-to-server-comm.png)
+
+The following sections break down these high-level steps in more detail.
+
+## Obtaining a token
+
+Before making a call to a resource server, the calling client first needs to [obtain a token](/oauth/token/).
+For this, a `client_credentials` grant is made using the caller's client ID and secret. Additionally,
+the caller should specify the resource (a.k.a. audience) trying to be accessed via the `resource`
```
Comment (zamzterz) [inline resources/articles/s2s-auth.md]: Relation between resource and audience is not 1:1, so maybe remove "(a.k.a. audience)".

File: resources/articles/s2s-auth.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+-d 'grant_type=client_credentials' https://login.schibsted.com/oauth/token
+```
+
+For further details on resource indicators, please refer to
+[Resource Indicators for OAuth 2.0, draft 2](https://tools.ietf.org/html/draft-campbell-oauth-resource-indicators-02).
+
+***Note:*** currently the value passed as the resource needs to match the domain registered in SPiD's
+[self service](/mobile/selfservice/) tool for the resource server.
+
+The scopes passed will help the receiving service authorize the call, and can be thought of a set of
```
Comment (zamzterz) [inline resources/articles/s2s-auth.md]: Missing word: "can be thought of a set of" -> "can be thought of as a set of"

File: resources/articles/s2s-auth.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  "client_id": "4e8463569caf7ca019000007",
+  "token_type": "Bearer",
+  "exp": 1487169930,
+  "iat": 1487166577,
+  "iss": "https://login.schibsted.com",
+  "aud": ["https://service.schibsted.com/messages"]
+}
+```
+
+One the resource server has introspected the token, it's important to validate the claims as follows:
```
Comment (zamzterz) [inline resources/articles/s2s-auth.md]: Typo: "One" -> "Once"

File: resources/articles/s2s-auth.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+1. Has the token expired? This is determined by checking the `exp` claim of the token. When performing
```
Comment (zamzterz) [inline resources/articles/s2s-auth.md]: Missing new line before this to render list correctly.

File: resources/articles/s2s-auth.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+1. Is the token issued by the expected issuer? If the resource server is using https://login.schibsted.com as
+its identity provider, it should only trust tokens with that URL in the `iss` claim.
+1. Does the token have the expected audience (determined by inspecting the `aud` claim, which is an array)?
+This is used to check if the token was intended for this service or not.
+1. Does the token have the expected scopes. The `scope` property contains a space-separated list of
+the scopes carried by the token. For each resource/action provided by the service, the implementers
+should decide on a set of scopes that are required to perform that action and ensure that the token
+carries those scopes.
+
+If all of the above checks pass, the resource server should let the call through. If the token has expired,
```
Comment (zamzterz) [inline resources/articles/s2s-auth.md]: Maybe include that missing or incorrect Authorization header should also result in `401 Unauthorized`.

File: resources/articles/s2s-auth.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+the service should return a status of `401 Unauthorized` to indicate to the calling service that it's
+time to fetch a new token. If the issuer, audience or scopes are incorrect/insufficient, the service
+should return a `403 Forbidden` to indicate that the token was understood, but the permissions are not
+sufficient to complete the call.
+
+Often services support multiple clients/tenants, and the `client_id` parameter can be used to determine
```
Comment (zamzterz) [inline resources/articles/s2s-auth.md]: Awesome paragraph! 👍

File: resources/articles/s2s-auth.md (hunk 1 above)
Comment (ksimons) [reply resources/articles/s2s-auth.md]: Fixed

File: resources/articles/s2s-auth.md (hunk 3 above)
Comment (ksimons) [reply resources/articles/s2s-auth.md]: Fixed

File: resources/articles/s2s-auth.md (hunk 2 above)
Comment (ksimons) [reply resources/articles/s2s-auth.md]: Fixed

File: resources/articles/s2s-auth.md (hunk 4 above)
Comment (ksimons) [reply resources/articles/s2s-auth.md]: Fixed

File: resources/articles/s2s-auth.md (hunk 5 above)
Comment (ksimons) [reply resources/articles/s2s-auth.md]: Fixed

File: resources/articles/s2s-auth.md (hunk 6 above)
Comment (ksimons) [reply resources/articles/s2s-auth.md]: Fixed

File: resources/articles/s2s-auth.md (hunk 7 above)
Comment (ksimons) [reply resources/articles/s2s-auth.md]: Fixed

File: resources/articles/s2s-auth.md (hunk 8 above)
Comment (ksimons) [reply resources/articles/s2s-auth.md]: Fixed

File: resources/articles/s2s-auth.md (hunk 9 above)
Comment (ksimons) [reply resources/articles/s2s-auth.md]: Fixed

**Pull request merged**